microsite/src/main/tut/usage/queue.md

```diff
@@ -0,0 +1,152 @@
+  _ <- queue.take
+  _ <- f.join
+} yield ()
+```
+
+It is also possible to add multiple values at once with `offerAll`:
+```tut:silent
+val res: IO[Nothing, Unit] = for {
+  queue <- Queue.bounded[Int](100)
+  items = Range.inclusive(1, 10).toList
+  _ <- queue.offerAll(items)
+} yield ()
+```
+
+## Consuming items from a queue
+
+`take` removes the oldest item from the queue and returns it. If the queue is empty, this will return a computation that resumes when an item has been added to the queue: you can use `fork` to wait for an item in a different fiber.
+```tut:silent
+val res: IO[Nothing, String] = for {
+  queue <- Queue.bounded[String](100)
+  f <- queue.take.fork // will be suspended because the queue is empty
+  _ <- queue.offer("something")
+  v <- f.join
+} yield v
+```
+
+You can consume multiple items at once with `takeUpTo`. If the queue doesn't have enough items to return, it will return all the items without waiting for more offers.
+```tut:silent
+val res: IO[Nothing, List[Int]] = for {
+  queue <- Queue.bounded[Int](100)
+  _ <- queue.offer(10)
+  _ <- queue.offer(20)
+  list  <- queue.takeUpTo(5)
+} yield list
+```
+
+Similarly, you can get all items at once with `takeAll`. It also returns without waiting (an empty list if the queue is empty).
+```tut:silent
+val res: IO[Nothing, List[Int]] = for {
+  queue <- Queue.bounded[Int](100)
+  _ <- queue.offer(10)
+  _ <- queue.offer(20)
+  list  <- queue.takeAll
+} yield list
+```
+
+## Shutting down a queue
+
+It is possible with `shutdown` to interrupt all the fibers that are suspended on `offer*` or `take*`. It will also empty the queue and make all future calls to `offer*` and `take*` terminate immediately.
+
+```tut:silent
+val res: IO[Nothing, Unit] = for {
+  queue <- Queue.bounded[Int](3)
+  f <- queue.take.fork
+  _ <- queue.shutdown // will interrupt f
+  _ <- f.join // will throw
+} yield ()
+```
+
+You can optionally pass a list of `Throwable` to `shutdown` so that they are given back as interruption `causes` to the failing fibers.
+
+```tut:silent
+val res: IO[Nothing, Unit] = for {
+  queue <- Queue.bounded[Int](3)
+  f <- queue.take.fork
+  _ <- queue.shutdown(new Exception("fail1"), new Exception("fail2")) // will interrupt f with those 2 exceptions
+  _ <- f.join // will throw
+} yield ()
+```
+
+You can use `onShutdown` to register a hook that will be called when the queue is shut down. It is possible to register multiple hooks.
+```tut:silent
+val res: IO[Nothing, Unit] = for {
+  queue <- Queue.bounded[Int](3)
+  p <- Promise.make[Nothing, Boolean]
+  _ <- queue.onShutdown(p.complete(true).void)
+  _ <- queue.shutdown
+  res <- p.get // will return true
+} yield ()
+```
+
+## Additional resources and examples
+
+- [ZIO Queue Talk by John De Goes @ ScalaWave 2018](https://www.slideshare.net/jdegoes/zio-queue)
+- [Elevator Control System using ZIO](https://medium.com/@wiemzin/elevator-control-system-using-zio-c718ae423c58)
```

Comment:
@ghostdogpr Very glad to see that! Thanks :)

Reply:
@ghostdogpr you can add this link if you found it useful:
 https://www.slideshare.net/wiemzin/psug-zio-queue

Reply:
Added!

Reply:
Cool :) Thanks :)